invisible_cities/detector_simulation/buffer_functions.py

```diff
@@ -0,0 +1,201 @@
+    sipm_buffer_samples = int(buffer_len * units.mus / sipm_binwid)
+    sipm_pretrg         = int(pre_trigger * units.mus / sipm_binwid)
+    sipm_postrg         = sipm_buffer_samples - sipm_pretrg
+    pmt_pretrg_         = int(pre_trigger * units.mus / pmt_binwid)
+    pmt_postrg_         = pmt_buffer_samples - pmt_pretrg_
+
+
+    def sipm_trg_bin(sipm_bins: np.ndarray,
+                     pmt_bins : np.ndarray) -> Callable[[int], int]:
+        """
+        Identifies the SiPM bin in which the trigger
+        was identified in the PMTs.
+        Assumes finer or equal bin size for PMTs compared to SiPMs
+        """
+        def get_sipm_bin(trigger: int) -> int:
+            return np.where(sipm_bins <= pmt_bins[trigger])[0][-1]
+        return get_sipm_bin
+
+
+    def slice_generator(pmt_bins   : np.ndarray,
+                        pmt_charge : np.ndarray,
+                        sipm_bins  : np.ndarray,
+                        sipm_charge: np.ndarray) -> Generator:
+        """
+        Synchronises the clocks between the SiPMs and PMTs
+        and slices the histograms where this synchronisation
+        indicates.
+        """
+        npmt_bin  = len(pmt_bins)
+        nsipm_bin = len(sipm_bins)
+        sipm_trg  = sipm_trg_bin(sipm_bins, pmt_bins)
+
+        def generate_slices(triggers: List) -> Tuple:
+
+            for trg in triggers:
+                trg_bin    = sipm_trg(trg)
+
+                bin_corr   = (pmt_bins[trg] - sipm_bins[trg_bin]) / pmt_binwid
+                pmt_pretrg = pmt_pretrg_ + int(bin_corr)
+                pmt_postrg = pmt_postrg_ - int(bin_corr)
+
+                pmt_pre    = 0       , trg - pmt_pretrg
+                pmt_pos    = npmt_bin, trg + pmt_postrg
+                pmt_sl     = slice(max(pmt_pre), min(pmt_pos))
+                pmt_pad    = (int(-min(pmt_pre)),
+                              int( max(0, pmt_pos[1] - npmt_bin)))
+
+                sipm_pre   = 0        , trg_bin - sipm_pretrg
+                sipm_pos   = nsipm_bin, trg_bin + sipm_postrg
+                sipm_sl    = slice(max(sipm_pre), min(sipm_pos))
+                sipm_pad   = (int(-min(sipm_pre)),
+                              int( max(0, sipm_pos[1] - nsipm_bin)))
+
+                yield ((pmt_charge [:,  pmt_sl],  pmt_pad),
+                       (sipm_charge[:, sipm_sl], sipm_pad))
+        return generate_slices
+
+
+    def position_signal(triggers   :       List,
+                        pmt_bins   : np.ndarray,
+                        pmt_charge :  pd.Series,
+                        sipm_bins  : np.ndarray,
+                        sipm_charge:  pd.Series) -> List:
+        """
+        Synchronises the SiPMs and PMTs for each identified
+        trigger and calls the padding function to fill with
+        zeros where necessary.
+        """
+
+        sipm_q = np.empty((0,0))\
+          if sipm_charge.empty else np.array(sipm_charge.tolist())
```

Comment:
Isn't `np.array(sipm_charge.tolist())` the same as `sipm_charge.values` or do you need to make a copy?
Why is the shape `(0,0)` if the series is empty? shouldn't it be `(0,)`?
If `(0,)` is the correct shape, then `sipm_q = sipm_charge.values` would cover both cases.

Reply:
Using `(0,0)` protects agains the use of 2D slicing in the other functions in the rare cases where we don't have SiPM info.